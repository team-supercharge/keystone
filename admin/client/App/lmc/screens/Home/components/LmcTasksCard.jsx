import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { Link } from 'react-router';
import {
    Button,
	GlyphButton,
	ResponsiveText,
} from '../../../../elemental';

const TaskCounter = (count, key) => {

    // http://localhost:3000/admin/tasks?filters=%5B%7B%22path%22%3A%22date%22%2C%22mode%22%3A%22before%22%2C%22inverted%22%3Afalse%2C%22value%22%3A%222018-04-09T11%3A00%3A00.000Z%22%2C%22before%22%3A%222018-04-09T00%3A00%3A00%2B01%3A00%22%2C%22after%22%3A%222018-04-09T00%3A00%3A00%2B01%3A00%22%7D%5D
    // need to ensure links numbers matched the link result! eg. filter by day?
    // only show these statuses
    const taskMap = {
        overdue: {
            color: "#e65d79e8",
            order: 1,
            params: '',   
        },
        pending: {
            color: "#ffba66",
            order: 2,
            params: 'filters=%5B%7B"path"%3A"status"%2C"inverted"%3Afalse%2C"value"%3A%5B"pending"%5D%7D%5D',   
        },
        completed: {
            color: "#9bd687e8",
            order: 3,
            params: 'filters=%5B%7B"path"%3A"status"%2C"inverted"%3Afalse%2C"value"%3A%5B"completed"%2C"skipped"%5D%7D%5D',   
        },
    };

    const background = {
        ...styles.taskCounter,
        background: taskMap[key] ? taskMap[key].color : '#e65d79',
    };

    const title = 'test';
    const to = `${Keystone.adminPath}/tasks?${taskMap[key].params}`;
    return (
        <Link
            key={key}
            tabIndex="-1"
            title={'test'}
            to={to}>
            <div style={background} className="">
                <strong>{ count }</strong> { key } today
            </div>
        </Link>
    )
}


class LmcTasksCard extends Component {
    render() {
        const { tasks, logs } = this.props;
        const logCountToday = logs && logs.length ? logs.length : 'No';
        
        const onClick = () => {
            this.props.onCreate('Task');
        }

        return (
            <div>
                <h2 className="lmc-card-title">
                    To-Do's & Logs
                </h2>
                <div className="lmc-card">
                    <div className="lmc-card-body">
                        {
                            tasks && tasks.results ? 
                                _.map(tasks.results, TaskCounter) : null
                        }
                        
                        
                    </div>
                    <div className="lmc-card-footer">
                        <div style={{ maxWidth: 145 }}>
                            <GlyphButton
                                block
                                color="success"
                                glyph="plus"
                                onClick={onClick}
                                position="left"
                                title={`Create To-Do`}
                            >
                                <ResponsiveText
                                    visibleSM="Create"
                                    visibleMD="Create"
                                    visibleLG={`Create To-Do`}
                                />
                            </GlyphButton>
                        </div>
                        <p style={{ marginBottom: 0 }}>{ logCountToday } logs recorded today</p>
                    </div>
                </div>
            </div>
        );
    }
}

const styles = {
    title: {
        opacity: 0.8,
    },
    taskCounter: {
        padding: '5px 10px',
        margin: 15,
        borderRadius: 30,
        fontWeight: 300,
        letterSpacing: 2,
        textTransform: 'uppercase',
        color: 'white',
    }
}

LmcTasksCard.propTypes = {

};

export default LmcTasksCard;
