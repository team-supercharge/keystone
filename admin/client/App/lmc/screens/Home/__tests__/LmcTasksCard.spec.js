import React from 'react'
import { shallow } from 'enzyme'
import MockDate from 'mockdate'
import LmcTasksCard from '../components/LmcTasksCard'

MockDate.set('1/1/2019')

describe('LmcTasksCard', () => {
    let wrapper
    let savedKeystone
    let onCreate
    let logs
    let tasks

    beforeAll(() => {
        savedKeystone = global.Keystone
        global.Keystone = { adminPath: 'TestPath' }
        onCreate = jest.fn()
    })

    beforeEach(() => {
        logs = [{}, {}, {}]
        tasks = {
            completed: 1,
            overdue: 1,
            pending: 1
        }

        wrapper = shallow(
            <LmcTasksCard
                logs={logs}
                tasks={tasks}
                onCreate={onCreate}
            />
        )
    })

    test('should render correct snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('onCreate has correct functionality', () => {
        wrapper.find('GlyphButton').simulate('click')
        expect(onCreate.mock.calls.length).toBe(1)
    })

    afterAll(() => {
        global.Keystone = savedKeystone
        MockDate.reset()
    })
})