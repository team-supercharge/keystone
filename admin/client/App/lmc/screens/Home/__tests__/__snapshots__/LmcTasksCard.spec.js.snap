// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LmcTasksCard should render correct snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <h2
      className="lmc-card-title"
    >
      To-Do's & Logs
    </h2>
    <div
      className="lmc-card"
    >
      <div
        className="lmc-card-body"
      >
        <div
          id="intro-js-step-tasks-info"
          style={
            Object {
              "paddingTop": 6,
            }
          }
        >
          <div
            style={
              Object {
                "paddingBottom": 20,
              }
            }
          >
            <Link
              onlyActiveOnIndex={false}
              style={
                Object {
                  "background": "#e65d79",
                  "borderRadius": 30,
                  "color": "white",
                  "fontWeight": 300,
                  "letterSpacing": 2,
                  "opacity": 0.9,
                  "padding": "5px 10px",
                  "textTransform": "uppercase",
                }
              }
              title="row.label Tasks"
              to="TestPath/tasks?filters=[{\\"path\\":\\"status\\",\\"value\\":[\\"pending\\"]},{\\"path\\":\\"date\\",\\"mode\\":\\"on\\",\\"value\\":\\"2019-01-01T12:00:00.000Z\\",\\"before\\":\\"2019-01-01T23:59:59.999Z\\",\\"after\\":\\"2019-01-01T00:00:00.000Z\\"}]"
            >
              <strong>
                1
              </strong>
               
              Overdue Today
            </Link>
            <br />
          </div>
          <div
            style={
              Object {
                "paddingBottom": 20,
              }
            }
          >
            <Link
              onlyActiveOnIndex={false}
              style={
                Object {
                  "background": "#ffba66",
                  "borderRadius": 30,
                  "color": "white",
                  "fontWeight": 300,
                  "letterSpacing": 2,
                  "opacity": 0.9,
                  "padding": "5px 10px",
                  "textTransform": "uppercase",
                }
              }
              title="row.label Tasks"
              to="TestPath/tasks?filters=[{\\"path\\":\\"status\\",\\"value\\":[\\"pending\\"]},{\\"path\\":\\"date\\",\\"mode\\":\\"on\\",\\"value\\":\\"2019-01-01T12:00:00.000Z\\",\\"before\\":\\"2019-01-01T23:59:59.999Z\\",\\"after\\":\\"2019-01-01T00:00:00.000Z\\"}]"
            >
              <strong>
                1
              </strong>
               
              Pending Today
            </Link>
            <br />
          </div>
          <div
            style={
              Object {
                "paddingBottom": 20,
              }
            }
          >
            <Link
              onlyActiveOnIndex={false}
              style={
                Object {
                  "background": "#b4d78b",
                  "borderRadius": 30,
                  "color": "white",
                  "fontWeight": 300,
                  "letterSpacing": 2,
                  "opacity": 0.9,
                  "padding": "5px 10px",
                  "textTransform": "uppercase",
                }
              }
              title="row.label Tasks"
              to="TestPath/tasks?filters=[{\\"path\\":\\"status\\",\\"value\\":[\\"completed\\",\\"skipped\\"]},{\\"path\\":\\"date\\",\\"mode\\":\\"on\\",\\"value\\":\\"2019-01-01T12:00:00.000Z\\",\\"before\\":\\"2019-01-01T23:59:59.999Z\\",\\"after\\":\\"2019-01-01T00:00:00.000Z\\"}]"
            >
              <strong>
                1
              </strong>
               
              Completed Today
            </Link>
            <br />
          </div>
        </div>
      </div>
      <div
        className="lmc-card-footer"
      >
        <div
          className="lmc-flex-container"
        >
          <div
            style={
              Object {
                "maxWidth": 145,
              }
            }
          >
            <GlyphButton
              block={true}
              color="success"
              glyph="plus"
              glyphStyle={Object {}}
              onClick={[Function]}
              position="left"
              title="Add To-Do"
            >
              <ResponsiveText
                component="span"
                visibleLG="Add To-Do"
                visibleMD="Add"
                visibleSM="Add"
              />
            </GlyphButton>
          </div>
          <p
            style={
              Object {
                "marginBottom": 0,
              }
            }
          >
            3
             
            logs
             recorded today
          </p>
        </div>
      </div>
    </div>
  </div>,
  "nodes": Array [
    <div>
      <h2
        className="lmc-card-title"
      >
        To-Do's & Logs
      </h2>
      <div
        className="lmc-card"
      >
        <div
          className="lmc-card-body"
        >
          <div
            id="intro-js-step-tasks-info"
            style={
              Object {
                "paddingTop": 6,
              }
            }
          >
            <div
              style={
                Object {
                  "paddingBottom": 20,
                }
              }
            >
              <Link
                onlyActiveOnIndex={false}
                style={
                  Object {
                    "background": "#e65d79",
                    "borderRadius": 30,
                    "color": "white",
                    "fontWeight": 300,
                    "letterSpacing": 2,
                    "opacity": 0.9,
                    "padding": "5px 10px",
                    "textTransform": "uppercase",
                  }
                }
                title="row.label Tasks"
                to="TestPath/tasks?filters=[{\\"path\\":\\"status\\",\\"value\\":[\\"pending\\"]},{\\"path\\":\\"date\\",\\"mode\\":\\"on\\",\\"value\\":\\"2019-01-01T12:00:00.000Z\\",\\"before\\":\\"2019-01-01T23:59:59.999Z\\",\\"after\\":\\"2019-01-01T00:00:00.000Z\\"}]"
              >
                <strong>
                  1
                </strong>
                 
                Overdue Today
              </Link>
              <br />
            </div>
            <div
              style={
                Object {
                  "paddingBottom": 20,
                }
              }
            >
              <Link
                onlyActiveOnIndex={false}
                style={
                  Object {
                    "background": "#ffba66",
                    "borderRadius": 30,
                    "color": "white",
                    "fontWeight": 300,
                    "letterSpacing": 2,
                    "opacity": 0.9,
                    "padding": "5px 10px",
                    "textTransform": "uppercase",
                  }
                }
                title="row.label Tasks"
                to="TestPath/tasks?filters=[{\\"path\\":\\"status\\",\\"value\\":[\\"pending\\"]},{\\"path\\":\\"date\\",\\"mode\\":\\"on\\",\\"value\\":\\"2019-01-01T12:00:00.000Z\\",\\"before\\":\\"2019-01-01T23:59:59.999Z\\",\\"after\\":\\"2019-01-01T00:00:00.000Z\\"}]"
              >
                <strong>
                  1
                </strong>
                 
                Pending Today
              </Link>
              <br />
            </div>
            <div
              style={
                Object {
                  "paddingBottom": 20,
                }
              }
            >
              <Link
                onlyActiveOnIndex={false}
                style={
                  Object {
                    "background": "#b4d78b",
                    "borderRadius": 30,
                    "color": "white",
                    "fontWeight": 300,
                    "letterSpacing": 2,
                    "opacity": 0.9,
                    "padding": "5px 10px",
                    "textTransform": "uppercase",
                  }
                }
                title="row.label Tasks"
                to="TestPath/tasks?filters=[{\\"path\\":\\"status\\",\\"value\\":[\\"completed\\",\\"skipped\\"]},{\\"path\\":\\"date\\",\\"mode\\":\\"on\\",\\"value\\":\\"2019-01-01T12:00:00.000Z\\",\\"before\\":\\"2019-01-01T23:59:59.999Z\\",\\"after\\":\\"2019-01-01T00:00:00.000Z\\"}]"
              >
                <strong>
                  1
                </strong>
                 
                Completed Today
              </Link>
              <br />
            </div>
          </div>
        </div>
        <div
          className="lmc-card-footer"
        >
          <div
            className="lmc-flex-container"
          >
            <div
              style={
                Object {
                  "maxWidth": 145,
                }
              }
            >
              <GlyphButton
                block={true}
                color="success"
                glyph="plus"
                glyphStyle={Object {}}
                onClick={[Function]}
                position="left"
                title="Add To-Do"
              >
                <ResponsiveText
                  component="span"
                  visibleLG="Add To-Do"
                  visibleMD="Add"
                  visibleSM="Add"
                />
              </GlyphButton>
            </div>
            <p
              style={
                Object {
                  "marginBottom": 0,
                }
              }
            >
              3
               
              logs
               recorded today
            </p>
          </div>
        </div>
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LmcTasksCard
        logs={
          Array [
            Object {},
            Object {},
            Object {},
          ]
        }
        onCreate={[MockFunction]}
        tasks={
          Object {
            "completed": 1,
            "overdue": 1,
            "pending": 1,
          }
        }
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LmcTasksCard {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "logs": Array [
            Object {},
            Object {},
            Object {},
          ],
          "onCreate": [MockFunction],
          "tasks": Object {
            "completed": 1,
            "overdue": 1,
            "pending": 1,
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <h2
            className="lmc-card-title"
          >
            To-Do's & Logs
          </h2>
          <div
            className="lmc-card"
          >
            <div
              className="lmc-card-body"
            >
              <div
                id="intro-js-step-tasks-info"
                style={
                  Object {
                    "paddingTop": 6,
                  }
                }
              >
                <div
                  style={
                    Object {
                      "paddingBottom": 20,
                    }
                  }
                >
                  <Link
                    onlyActiveOnIndex={false}
                    style={
                      Object {
                        "background": "#e65d79",
                        "borderRadius": 30,
                        "color": "white",
                        "fontWeight": 300,
                        "letterSpacing": 2,
                        "opacity": 0.9,
                        "padding": "5px 10px",
                        "textTransform": "uppercase",
                      }
                    }
                    title="row.label Tasks"
                    to="TestPath/tasks?filters=[{\\"path\\":\\"status\\",\\"value\\":[\\"pending\\"]},{\\"path\\":\\"date\\",\\"mode\\":\\"on\\",\\"value\\":\\"2019-01-01T12:00:00.000Z\\",\\"before\\":\\"2019-01-01T23:59:59.999Z\\",\\"after\\":\\"2019-01-01T00:00:00.000Z\\"}]"
                  >
                    <strong>
                      1
                    </strong>
                     
                    Overdue Today
                  </Link>
                  <br />
                </div>
                <div
                  style={
                    Object {
                      "paddingBottom": 20,
                    }
                  }
                >
                  <Link
                    onlyActiveOnIndex={false}
                    style={
                      Object {
                        "background": "#ffba66",
                        "borderRadius": 30,
                        "color": "white",
                        "fontWeight": 300,
                        "letterSpacing": 2,
                        "opacity": 0.9,
                        "padding": "5px 10px",
                        "textTransform": "uppercase",
                      }
                    }
                    title="row.label Tasks"
                    to="TestPath/tasks?filters=[{\\"path\\":\\"status\\",\\"value\\":[\\"pending\\"]},{\\"path\\":\\"date\\",\\"mode\\":\\"on\\",\\"value\\":\\"2019-01-01T12:00:00.000Z\\",\\"before\\":\\"2019-01-01T23:59:59.999Z\\",\\"after\\":\\"2019-01-01T00:00:00.000Z\\"}]"
                  >
                    <strong>
                      1
                    </strong>
                     
                    Pending Today
                  </Link>
                  <br />
                </div>
                <div
                  style={
                    Object {
                      "paddingBottom": 20,
                    }
                  }
                >
                  <Link
                    onlyActiveOnIndex={false}
                    style={
                      Object {
                        "background": "#b4d78b",
                        "borderRadius": 30,
                        "color": "white",
                        "fontWeight": 300,
                        "letterSpacing": 2,
                        "opacity": 0.9,
                        "padding": "5px 10px",
                        "textTransform": "uppercase",
                      }
                    }
                    title="row.label Tasks"
                    to="TestPath/tasks?filters=[{\\"path\\":\\"status\\",\\"value\\":[\\"completed\\",\\"skipped\\"]},{\\"path\\":\\"date\\",\\"mode\\":\\"on\\",\\"value\\":\\"2019-01-01T12:00:00.000Z\\",\\"before\\":\\"2019-01-01T23:59:59.999Z\\",\\"after\\":\\"2019-01-01T00:00:00.000Z\\"}]"
                  >
                    <strong>
                      1
                    </strong>
                     
                    Completed Today
                  </Link>
                  <br />
                </div>
              </div>
            </div>
            <div
              className="lmc-card-footer"
            >
              <div
                className="lmc-flex-container"
              >
                <div
                  style={
                    Object {
                      "maxWidth": 145,
                    }
                  }
                >
                  <GlyphButton
                    block={true}
                    color="success"
                    glyph="plus"
                    glyphStyle={Object {}}
                    onClick={[Function]}
                    position="left"
                    title="Add To-Do"
                  >
                    <ResponsiveText
                      component="span"
                      visibleLG="Add To-Do"
                      visibleMD="Add"
                      visibleSM="Add"
                    />
                  </GlyphButton>
                </div>
                <p
                  style={
                    Object {
                      "marginBottom": 0,
                    }
                  }
                >
                  3
                   
                  logs
                   recorded today
                </p>
              </div>
            </div>
          </div>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <h2
            className="lmc-card-title"
          >
            To-Do's & Logs
          </h2>
          <div
            className="lmc-card"
          >
            <div
              className="lmc-card-body"
            >
              <div
                id="intro-js-step-tasks-info"
                style={
                  Object {
                    "paddingTop": 6,
                  }
                }
              >
                <div
                  style={
                    Object {
                      "paddingBottom": 20,
                    }
                  }
                >
                  <Link
                    onlyActiveOnIndex={false}
                    style={
                      Object {
                        "background": "#e65d79",
                        "borderRadius": 30,
                        "color": "white",
                        "fontWeight": 300,
                        "letterSpacing": 2,
                        "opacity": 0.9,
                        "padding": "5px 10px",
                        "textTransform": "uppercase",
                      }
                    }
                    title="row.label Tasks"
                    to="TestPath/tasks?filters=[{\\"path\\":\\"status\\",\\"value\\":[\\"pending\\"]},{\\"path\\":\\"date\\",\\"mode\\":\\"on\\",\\"value\\":\\"2019-01-01T12:00:00.000Z\\",\\"before\\":\\"2019-01-01T23:59:59.999Z\\",\\"after\\":\\"2019-01-01T00:00:00.000Z\\"}]"
                  >
                    <strong>
                      1
                    </strong>
                     
                    Overdue Today
                  </Link>
                  <br />
                </div>
                <div
                  style={
                    Object {
                      "paddingBottom": 20,
                    }
                  }
                >
                  <Link
                    onlyActiveOnIndex={false}
                    style={
                      Object {
                        "background": "#ffba66",
                        "borderRadius": 30,
                        "color": "white",
                        "fontWeight": 300,
                        "letterSpacing": 2,
                        "opacity": 0.9,
                        "padding": "5px 10px",
                        "textTransform": "uppercase",
                      }
                    }
                    title="row.label Tasks"
                    to="TestPath/tasks?filters=[{\\"path\\":\\"status\\",\\"value\\":[\\"pending\\"]},{\\"path\\":\\"date\\",\\"mode\\":\\"on\\",\\"value\\":\\"2019-01-01T12:00:00.000Z\\",\\"before\\":\\"2019-01-01T23:59:59.999Z\\",\\"after\\":\\"2019-01-01T00:00:00.000Z\\"}]"
                  >
                    <strong>
                      1
                    </strong>
                     
                    Pending Today
                  </Link>
                  <br />
                </div>
                <div
                  style={
                    Object {
                      "paddingBottom": 20,
                    }
                  }
                >
                  <Link
                    onlyActiveOnIndex={false}
                    style={
                      Object {
                        "background": "#b4d78b",
                        "borderRadius": 30,
                        "color": "white",
                        "fontWeight": 300,
                        "letterSpacing": 2,
                        "opacity": 0.9,
                        "padding": "5px 10px",
                        "textTransform": "uppercase",
                      }
                    }
                    title="row.label Tasks"
                    to="TestPath/tasks?filters=[{\\"path\\":\\"status\\",\\"value\\":[\\"completed\\",\\"skipped\\"]},{\\"path\\":\\"date\\",\\"mode\\":\\"on\\",\\"value\\":\\"2019-01-01T12:00:00.000Z\\",\\"before\\":\\"2019-01-01T23:59:59.999Z\\",\\"after\\":\\"2019-01-01T00:00:00.000Z\\"}]"
                  >
                    <strong>
                      1
                    </strong>
                     
                    Completed Today
                  </Link>
                  <br />
                </div>
              </div>
            </div>
            <div
              className="lmc-card-footer"
            >
              <div
                className="lmc-flex-container"
              >
                <div
                  style={
                    Object {
                      "maxWidth": 145,
                    }
                  }
                >
                  <GlyphButton
                    block={true}
                    color="success"
                    glyph="plus"
                    glyphStyle={Object {}}
                    onClick={[Function]}
                    position="left"
                    title="Add To-Do"
                  >
                    <ResponsiveText
                      component="span"
                      visibleLG="Add To-Do"
                      visibleMD="Add"
                      visibleSM="Add"
                    />
                  </GlyphButton>
                </div>
                <p
                  style={
                    Object {
                      "marginBottom": 0,
                    }
                  }
                >
                  3
                   
                  logs
                   recorded today
                </p>
              </div>
            </div>
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LmcTasksCard
    logs={
      Array [
        Object {},
        Object {},
        Object {},
      ]
    }
    onCreate={[MockFunction]}
    tasks={
      Object {
        "completed": 1,
        "overdue": 1,
        "pending": 1,
      }
    }
  />,
}
`;
